### ISSUE TITLE
Using Triton as a custom service in GitLab CI 

### ISSUE THREAD

**Issue by elmuz** [triton-inference-server/server#3292]
I think some of you already use Triton as part of your CI. I want to do the same but I am facing some issues to make the process nice and clean. Basically, I do not know how to push models to Triton before the main job is executed. Ideally, models are part of the main repository and pushed to Triton model registry dynamically at every job creation, but this creates a synchronization issue: all services (i.e. Triton) are spawn before the main environment, but Triton wants a working model registry or it goes into a broken state.

From the documentation we see that 2 general approaches can be used:
- *Local storage*. This is fine, but it means that Triton should implement a kind of FTP service  (e.g. in `/models`) so that the main actor can push there models before running the tests. Alternatively, in this fashion, I managed to use a custom Triton container, created dynamically by the CI during a previous stage, where its dockerfile has a `COPY` instructions for the models. This is not ideal since you have a Docker build and push instruction at each  new experiment.
- *Remote storage*. In this way we should mock some S3/GCP/... bucket and link it to Triton. This is tricky. First of all you want the bucket to be spawn before Triton, otherwise it crashes (not really sure if GitLab spawn services in parallel or not). I tried with S3 mock (`motoserver`) but it wasn't working, since the S3 does not have a bucket as it starts. Buckets are created from clients (i.e. the main service, which is not started yet) and Triton crashes if not finding it. So some tricky synchronization solution must be set.

Do you have any experience on this regard? What's your approach?

**Comment by dyastremsky** [comment]
@elmuz Have you considered mounting a Docker volume with the model repository? That seems to achieve the local storage case you outlined without needing to rebuild the container.

**Comment by elmuz** [comment]
@dyastremsky IIRC GitLab CI yaml definition does not have volume configuration for services. Even if I set `VOLUME` in Triton Dockerfile, how can I access it from the main environment? Can you give me an example?

**Comment by dyastremsky** [comment]
I'm less familiar with GitLab CI specifics, so tagging @dzier who may be able to provide further guidance.

**Comment by dzier** [comment]
@elmuz for our internal CI, we have a cluster of Gitlab runner machines that all mount to a shared NSF drive that contains all of our models.  In the gitlab-ci.yml file, we just create a variable that points to the base directory, for example:

```
variables:
   DATADIR: /mnt/models
```

then in for our tests, we create a test template with the `docker run` command that maps in the directory with `-v $DATADIR:/data`. All of our scripts are ran inside the docker container and can access the models locally at `/data`. You can see many examples of this is used in the `test.sh` files within any of the qa/ci tests we have in https://github.com/triton-inference-server/server/tree/main/qa.

For examples of how we test remote storage, please refer to https://github.com/triton-inference-server/server/blob/main/qa/L0_storage_S3/test.sh.

**Comment by elmuz** [comment]
@dzier thank you very much for your exhaustive reply. In the meanwhile I was able to come up with a solution that fits our need. I will share here in case anyone else wants to take some ideas from it.

As a data scientist I develop models and I want my code to be decoupled from inference logic. Also I want Triton CI service to be as much vanilla as possible at every CI execution. I am quite against having a shared folder as a model repository, since there's the risk of garbage silently growing and becoming soon a real mess. During experiments new models are generated frequently and I do not want other pipelines to be affected by unwanted overwriting. Moreover, I want that all the garbage to live in the time-space of the CI job and nothing more (unless promoted further). GitLab has some limitation compared to docker-compose configurations, but nevertheless something working is achievable.

I found that a CI job of the above kind (for test/evaluation) can be done with the help of two auxiliary services: Triton of course and an S3 mock (e.g. [minio](https://docs.min.io/)). Here's some key facts:
- Triton is spawn using an S3 bucket as a model registry. At the beginning of time the bucket is empty. Minio container can be started with an already existing S3 bucket using `bin/sh` as _entrypoint_ and ` ["-c", "mkdir -p /buckets/models && /usr/bin/minio server /buckets"]` as _command_ . Obviously, access key/psw must be consistent between the two services.
- Here there's a caveat: Triton crashes if S3 is not reachable (yet) for some reasons. I couldn't find a way to force execution order here. However, Triton is definitely slower to spin up and always comes after S3. In case of troubles a "wait-for.sh" has to be injected in Triton's _command_. Optionally, setting the model repository location via API could be a nice alternative (is it possible?).
- in the `before_script` instruction I run a Python script that pushes all the models (the artifacts in the main repo) to the S3 container one after the other. Triton APIs are also used to explicitly load each models after they are pushed.
- Triton `--model-control-mode` is set to `explicit`. This means that there is not need to wait for a polling interval. It's the client that activates model and doesn't need to wait for the server polling routine. This eliminates one of the synchronization issues I was facing.

I had to tweak here and there but in the end it is quite robust, controllable and easy to maintain across several company-projects. I will close this for now since personally I solved my issues for the moment.
